cpp/src/community/detail/refine.cuh

```diff
@@ -0,0 +1,763 @@
+    handle,
+    graph_view,
+    GraphViewType::is_multi_gpu
+      ? src_louvain_assignment_cache.view()
+      : detail::edge_major_property_view_t<vertex_t, vertex_t const*>(louvain_assignment.data()),
+    GraphViewType::is_multi_gpu ? dst_louvain_assignment_cache.view()
+                                : detail::edge_minor_property_view_t<vertex_t, vertex_t const*>(
+                                    louvain_assignment.data(), vertex_t{0}),
+    *edge_weight_view,
+    [] __device__(auto src, auto dst, auto src_cluster, auto dst_cluster, auto wt) {
+      weight_t weighted_deg_contribution{wt};
+      weight_t weighted_cut_contribution{0};
+
+      if (src == dst)  // self loop
+        weighted_cut_contribution = 0;
+      else if (src_cluster == dst_cluster)
+        weighted_cut_contribution = wt;
+
+      return thrust::make_tuple(weighted_deg_contribution, weighted_cut_contribution);
+    },
+    thrust::make_tuple(weight_t{0}, weight_t{0}),
+    cugraph::reduce_op::plus<thrust::tuple<weight_t, weight_t>>{},
+    thrust::make_zip_iterator(thrust::make_tuple(weighted_degree_of_vertices.begin(),
+                                                 weighted_cut_of_vertices_to_louvain.begin())));
+
+  rmm::device_uvector<uint8_t> singleton_and_connected_flags(
+    graph_view.local_vertex_partition_range_size(), handle.get_stream());
+
+  auto wcut_deg_and_cluster_vol_triple_begin =
+    thrust::make_zip_iterator(thrust::make_tuple(weighted_cut_of_vertices_to_louvain.begin(),
+                                                 weighted_degree_of_vertices.begin(),
+                                                 vertex_cluster_weights_v.begin()));
+  auto wcut_deg_and_cluster_vol_triple_end =
+    thrust::make_zip_iterator(thrust::make_tuple(weighted_cut_of_vertices_to_louvain.end(),
+                                                 weighted_degree_of_vertices.end(),
+                                                 vertex_cluster_weights_v.end()));
+
+  thrust::transform(handle.get_thrust_policy(),
+                    wcut_deg_and_cluster_vol_triple_begin,
+                    wcut_deg_and_cluster_vol_triple_end,
+                    singleton_and_connected_flags.begin(),
+                    [gamma = resolution] __device__(auto wcut_wdeg_and_louvain_volume) {
+                      auto wcut           = thrust::get<0>(wcut_wdeg_and_louvain_volume);
+                      auto wdeg           = thrust::get<1>(wcut_wdeg_and_louvain_volume);
+                      auto louvain_volume = thrust::get<2>(wcut_wdeg_and_louvain_volume);
+                      return wcut > (gamma * wdeg * (louvain_volume - wdeg));
+                    });
+
+  // Update cluster weight, weighted degree and cut for edge sources
+
+  edge_src_property_t<GraphViewType, weight_t> src_louvain_cluster_weight_cache(handle);
+  edge_src_property_t<GraphViewType, thrust::tuple<weight_t, weight_t>>
+    src_wdeg_and_cut_to_Louvain_cache(handle);
```

Comment:
And any reason to use uppercase "L" in "Louvain" here? (while using lowercase "l" in other places).